rfc/JENKINS-41334-parallel-stage-execution.md

```diff
@@ -0,0 +1,86 @@
+## RFC: [JENKINS-41334](https://issues.jenkins-ci.org/browse/JENKINS-41334) - Parallel stage execution
+
+* Author: Andrew Bayer
+* Created Date: February 27, 2017
+* Approved Date: n/a
+* JIRA: [JENKINS-41334](https://issues.jenkins-ci.org/browse/JENKINS-41334)
+* Status: Draft
+
+### Problem Statement
+
+The `parallel` step does not fit smoothly into Declarative. It doesn't
+allow for parallel stage execution, it doesn't utilize the `agent`
+definition, etc. There needs to be the ability to run `stage`s in
+parallel, with the full power of the Declarative syntax enabled -
+i.e., `agent`, `post`, `when`, etc.
+
+### Syntax
+
+#### Proposed example
+
+```groovy
+stage('foo') {
+    parallel {
+        stage('first') {
+            steps {
+                echo "First branch"
+            }
+        }
+        stage('second') {
+            steps {
+                echo "Second branch"
+            }
+        }
+    }
+}
+```
+
+#### Details
+
+* A `stage` has to contain one and only one of `steps` or
+    `parallel`. They cannot be combined in the same `stage`.
+* `stage` configuration can be done on both a `stage` containing
+    `parallel`, or `stage`s within `parallel`, i.e., `agent`, `when`,
+    `post`, etc.
+* `agent`, `environment`, and `tools` specified on the "parent"
+    `stage` will apply to child `stage`s in the same way that top-level
+    configuration applies to `stage`s currently.
+* Naming of `parallel` is up for debate.
+
+### Runtime Implementation
+
+There are two options for runtime. Which one we go with depends on
+Blue Ocean visualization decisions. The first option is to simply nest
+stage executions, while the second option is to not actually use the
+`stage` step for the nested `stage`s and instead utilize the `stage`
+names as the names of parallel branches.
+
+These two options each have their pros and cons. A pro in one option
+tends to lead to a con in the other option.
+
+#### Pro/Con
+
+* Stage status/metadata (i.e., skipped due to failure/`when`) would
```

Comment:
@vivek What I mean is that if you want to use custom markers for status as a new feature, that would need support... perhaps "fixed" was entirely the wrong word for this, "enhanced" would be more accurate.

As it stands the use of the Visitor/callback pattern + graph-analysis library should work right, since it converts the forks into a series of linear start-to-stop segments with orderly rules for status/timing.  

If we decide to break from normal pipeline error/flow patterns for Declarative though, and use some other mechanism to denote pass/fail/not-run/etc then all bets are of course off until that is supported.